### PR TITLE
Add new cli configure options

### LINKED ISSUE #2295: [Enhancement] `dab configure --data-source` in CLI
Support the following options for `dab configure` per #2251

| Configuration File Property Path                                        | CLI Flag                                                           | Data Type | Nullable |
|-------------------------------------------------------------------------|--------------------------------------------------------------------|-----------|----------|
| data-source.<br/>database-type                                          | <nobr>--data-source.database-type</nobr>                           | String: `MSSQL`, `PostgreSQL`, `CosmosDB`, `MySQL` | ❌ |
| data-source.<br/>connection-string                                      | <nobr>--data-source.connection-string</nobr>                       | String    | ❌       |
| data-source.<br/>options.database                                       | <nobr>--data-source.options.database</nobr>                        | String    | ✅       |
| data-source.<br/>options.container                                      | <nobr>--data-source.options.container</nobr>                       | String    | ✅       |
| data-source.<br/>options.schema                                         | <nobr>--data-source.options.schema</nobr>                          | String    | ✅       |
| data-source.<br/>options.set-session-context                            | <nobr>--data-source.options.set-session-context</nobr>             | Boolean: `true`, `false` (default: `false`) | ✅ |

### PR DESCRIPTION
**Pull request by abhishekkumams**
## Why make this change?

- Closes #2295 
  - This change will allow user to update data-source options.
  - Also, it makes CLI more powerful and developer friendly.

## What is this change?

- Introducing new options for configure command to update data-source options.

| Configuration File Property Path                                        | CLI Flag                                                           | Data Type | Nullable |
|-------------------------------------------------------------------------|--------------------------------------------------------------------|-----------|----------|
| data-source.<br/>database-type                                          | <nobr>--data-source.database-type</nobr>                           | String: `MSSQL`, `PostgreSQL`, `CosmosDB_NoSQL`, `MySQL` | ✅  |
| data-source.<br/>connection-string                                      | <nobr>--data-source.connection-string</nobr>                       | String    | ✅        |
| data-source.<br/>options.database                                       | <nobr>--data-source.options.database</nobr>                        | String    | ✅       |
| data-source.<br/>options.container                                      | <nobr>--data-source.options.container</nobr>                       | String    | ✅       |
| data-source.<br/>options.schema                                         | <nobr>--data-source.options.schema</nobr>                          | String    | ✅       |
| data-source.<br/>options.set-session-context                            | <nobr>--data-source.options.set-session-context</nobr>             | Boolean: `true`, `false` (default: `false`) | ✅   |

## NOTES:
Certain properties that are specific to database can only be updated with the correct databaseType.
For Example: If the database type initially was MSSQL, and we give cosmosDBoptions without specifying the new dbType as cosmosDB_NoSQL, the command will fail.
For better usability, we will also allow users to first update the dbType to cosmosDB_NoSQL and then in a new command provide the cosmosDB specifig options, since using all the options in one command can be very long and errornous. 

## How was this tested?

- [x] End-End Tests
- [x] Unit Tests

## Sample Request(s)

- To update to mssql
`dab configure --data-source.database-type mssql --data-source.options.set-session-context true`

- To update to cosmosDB_NoSQL
   #### Single command
  `dab configure --data-source.database-type cosmosdb_nosql --data-source.options.database testdbname --data-source.options.schema testschema.gql`

  #### Break the command
  1. `dab configure --data-source.database-type cosmosdb_nosql`
  2. `dab configure --data-source.options.database testdbname --data-source.options.schema testschema.gql`

### CODE REVIEW

File: src/Cli.Tests/ConfigureOptionsTests.cs
Comment (seantleonard) [inline src/Cli.Tests/ConfigureOptionsTests.cs:1]: in pr description you have:
> For better usability, we will also allow users to first update the dbType to cosmosDB_NoSQL and then in a new command provide the cosmosDB specifig options, since using all the options in one command can be very long and errornous.

and then an example to break up the command: 
> 1. `dab configure --data-source.database-type cosmosdb_nosql`
> 2. `dab configure --data-source.options.database testdbname --data-source.options.schema testschema.gql`

So, user must remember to run the second command, otherwise their config is in a bad state? I'm on the fence about this because this implies that execution of a single DAB cli command *could* result in your config being invalid.

File: src/Cli/Commands/ConfigureOptions.cs [hunk 1]
```diff
@@ -19,13 +19,43 @@ namespace Cli.Commands
...
+        public string? DataSourceConnectionString { get; }
+
+        [Option("data-source.options.database", Required = false, HelpText = "Database name for Cosmos DB for NoSql.")]
+        public string? DataSourceOptionsDatabase { get; }
+
+        [Option("data-source.options.container", Required = false, HelpText = "Container name for Cosmos DB for NoSql.")]
+        public string? DataSourceOptionsContainer { get; }
+
+        [Option("data-source.options.schema", Required = false, HelpText = "Schema path for Cosmos DB for NoSql.")]
+        public string? DataSourceOptionsSchema { get; }
```
Comment (seantleonard) [inline src/Cli/Commands/ConfigureOptions.cs:54]: @sajeetharan, can you provide feedback on this help text for Cosmos? if i recall, we need to specify Azure Cosmos DB instead of Cosmos DB for NoSql?

File: src/Cli/Commands/ConfigureOptions.cs [hunk 2]
```diff
@@ -19,13 +19,43 @@ namespace Cli.Commands
...
+
+        [Option("data-source.options.set-session-context", Required = false, HelpText = "Set session context. Allowed values: true, false (default: false).")]
```
Comment (seantleonard) [inline src/Cli/Commands/ConfigureOptions.cs:56]:
```suggestion
        [Option("data-source.options.set-session-context", Required = false, HelpText = "Enable session context. Allowed values: true (default), false.")]
```

default is true per datasource.cs line 87

File: src/Cli/ConfigGenerator.cs [hunk 3]
```diff
@@ -535,6 +541,98 @@ public static bool TryConfigureSettings(ConfigureOptions options, FileSystemRunt
             return WriteRuntimeConfigToFile(runtimeConfigFile, runtimeConfig, fileSystem);
         }
 
+        /// <summary>
+        /// Configures the data source options for the runtimeconfig based on the provided options.
+        /// This method updates the database type, connection string, and other database-specific options in the config file.
+        /// It validates the provided database type and ensures that options specific to certain database types are correctly applied.
+        /// If any validation fails, it logs an error and returns false.
```
Comment (seantleonard) [inline src/Cli/ConfigGenerator.cs:548]:
```suggestion
        /// When validation fails, this function logs the validation errors and returns false.
```

File: src/Cli/ConfigGenerator.cs [hunk 4]
```diff
@@ -535,6 +541,98 @@ public static bool TryConfigureSettings(ConfigureOptions options, FileSystemRunt
...
+            {
+                AddCosmosDbOptions(dbOptions, options, namingPolicy);
+            }
+            else if (!string.IsNullOrWhiteSpace(options.DataSourceOptionsDatabase)
+                    || !string.IsNullOrWhiteSpace(options.DataSourceOptionsContainer)
+                    || !string.IsNullOrWhiteSpace(options.DataSourceOptionsSchema))
+            {
+                _logger.LogError("Database, Container, and Schema options are only applicable for CosmosDB_NoSQL database type.");
+                return false;
+            }
```
Comment (seantleonard) [inline src/Cli/ConfigGenerator.cs:589]: when changing dbtype from sql to cosmos, users will need to change and provid cosmos options in a single command, right? Maybe i haven't read far enough but it seems like the 2 step command approach is prevented by this condition.

File: src/Cli/ConfigGenerator.cs [hunk 5]
```diff
@@ -535,6 +541,98 @@ public static bool TryConfigureSettings(ConfigureOptions options, FileSystemRunt
...
+
+                dbOptions.Add(namingPolicy.ConvertName(nameof(MsSqlOptions.SetSessionContext)), options.DataSourceOptionsSetSessionContext.Value);
+            }
+
+            dbOptions = dbOptions.IsNullOrEmpty() ? null : dbOptions;
+            DataSource dataSource = new(dbType, dataSourceConnectionString, dbOptions);
+            runtimeConfig = runtimeConfig with { DataSource = dataSource };
+
+            return true;
+        }
```
Comment (seantleonard) [inline src/Cli/ConfigGenerator.cs:607]: Should returning true be conditioned on runtimeconfig != null to honor the promise you make in the function signature and description?

File: src/Cli/ConfigGenerator.cs [hunk 6]
```diff
@@ -12,6 +12,7 @@
 using Azure.DataApiBuilder.Service;
 using Cli.Commands;
 using Microsoft.Extensions.Logging;
+using Microsoft.IdentityModel.Tokens;
```
Comment (seantleonard) [inline src/Cli/ConfigGenerator.cs:15]: why do you need this? Did i miss code which needs this?

File: src/Cli.Tests/ConfigureOptionsTests.cs [hunk 7]
```diff
@@ -150,5 +150,195 @@ public void TestUpdateDepthLimitForGraphQL(int? newDepthLimit)
 
             Assert.AreEqual(newDepthLimit, config.Runtime?.GraphQL?.DepthLimit);
         }
+
+        /// <summary>
+        /// Tests the update of the database type in the runtime config.
```
Comment (seantleonard) [inline src/Cli.Tests/ConfigureOptionsTests.cs:155]:
```suggestion
        /// Tests running "dab config -update"
```
something like the above to demonstrate which cli command/usage is tested.

File: src/Cli.Tests/ConfigureOptionsTests.cs [hunk 8]
```diff
@@ -150,5 +150,195 @@ public void TestUpdateDepthLimitForGraphQL(int? newDepthLimit)
...
+
+            // Assert
+            string updatedConfig = _fileSystem!.File.ReadAllText(TEST_RUNTIME_CONFIG_FILE);
+            Assert.IsTrue(RuntimeConfigLoader.TryParseConfig(updatedConfig, out config));
+            Assert.IsNotNull(config.Runtime);
+            Assert.AreEqual(config.DataSource.DatabaseType, Enum.Parse<DatabaseType>(dbType, ignoreCase: true));
+        }
+
+        /// <summary>
+        /// Tests the update of the database type from CosmosDB_NoSQL to MSSQL in the runtime config.
```
Comment (seantleonard) [inline src/Cli.Tests/ConfigureOptionsTests.cs:191]: can you include which command(s) is used? one or two commands?

File: src/Cli.Tests/ConfigureOptionsTests.cs [hunk 9]
```diff
@@ -150,5 +150,195 @@ public void TestUpdateDepthLimitForGraphQL(int? newDepthLimit)
...
+        /// This method verifies that the database type can be changed from CosmosDB_NoSQL to MSSQL and that the 
+        /// specific MSSQL option 'set-session-context' is correctly added to the configuration and the specific
+        /// cosmosDB options are removed. 
+        /// </summary>
+        [TestMethod]
+        public void TestDatabaseTypeUpdateCosmosDB_NoSQLToMSSQL()
+        {
+            // Arrange
+            _fileSystem!.AddFile(TEST_RUNTIME_CONFIG_FILE, new MockFileData(INITIAL_COSMOSDB_NOSQL_CONFIG));
```
Comment (seantleonard) [inline src/Cli.Tests/ConfigureOptionsTests.cs:200]: does this overwrite a pre-existing file? i'd expect so to mitigate tests stepping over eachother.

File: src/Cli.Tests/ConfigureOptionsTests.cs [hunk 10]
```diff
@@ -150,5 +150,195 @@ public void TestUpdateDepthLimitForGraphQL(int? newDepthLimit)
...
+                config: TEST_RUNTIME_CONFIG_FILE
+            );
+            Assert.IsTrue(TryConfigureSettings(options, _runtimeConfigLoader!, _fileSystem!));
+
+            // Assert
+            string updatedConfig = _fileSystem!.File.ReadAllText(TEST_RUNTIME_CONFIG_FILE);
+            Assert.IsTrue(RuntimeConfigLoader.TryParseConfig(updatedConfig, out config));
+            Assert.IsNotNull(config.Runtime);
+            Assert.AreEqual(config.DataSource.DatabaseType, DatabaseType.MSSQL);
+            Assert.AreEqual(config.DataSource.Options!.GetValueOrDefault("set-session-context", false), true);
```
Comment (seantleonard) [inline src/Cli.Tests/ConfigureOptionsTests.cs:220]: what does the `false` value do here: `Options!.GetValueOrDefault("set-session-context", false), true);`

File: src/Cli.Tests/ConfigureOptionsTests.cs [hunk 11]
```diff
@@ -150,5 +150,195 @@ public void TestUpdateDepthLimitForGraphQL(int? newDepthLimit)
...
+            Assert.IsFalse(config.DataSource.Options!.ContainsKey("database"));
+            Assert.IsFalse(config.DataSource.Options!.ContainsKey("container"));
+            Assert.IsFalse(config.DataSource.Options!.ContainsKey("schema"));
+        }
+
+        /// <summary>
+        /// Tests the update of the database type from MSSQL to CosmosDB_NoSQL in the runtime config.
+        /// This method verifies that the database type can be changed from MSSQL to CosmosDB_NoSQL and that the 
+        /// specific CosmosDB_NoSQL options such as database, container, and schema are correctly added to the config.
```
Comment (seantleonard) [inline src/Cli.Tests/ConfigureOptionsTests.cs:229]: using which commands, and how many? 1, 2?

File: src/Cli.Tests/ConfigureOptionsTests.cs [hunk 12]
```diff
@@ -150,5 +150,195 @@ public void TestUpdateDepthLimitForGraphQL(int? newDepthLimit)
...
+            Assert.IsNotNull(config.Runtime);
+
+            // Act
+            ConfigureOptions options = new(
+                dataSourceDatabaseType: "invalid",
+                config: TEST_RUNTIME_CONFIG_FILE
+            );
+
+            // Assert
+            Assert.IsFalse(TryConfigureSettings(options, _runtimeConfigLoader!, _fileSystem!));
```
Comment (seantleonard) [inline src/Cli.Tests/ConfigureOptionsTests.cs:286]: The "act" the last line where `TryConfigureSettings` is used

File: src/Cli.Tests/ConfigureOptionsTests.cs [hunk 13]
```diff
@@ -150,5 +150,195 @@ public void TestUpdateDepthLimitForGraphQL(int? newDepthLimit)
...
+            // Act
+            ConfigureOptions options = new(
+                dataSourceOptionsDatabase: "testdb",
+                dataSourceOptionsContainer: "testcontainer",
+                dataSourceOptionsSchema: "testschema.gql",
+                config: TEST_RUNTIME_CONFIG_FILE
+            );
+
+            // Assert
+            Assert.IsFalse(TryConfigureSettings(options, _runtimeConfigLoader!, _fileSystem!));
```
Comment (seantleonard) [inline src/Cli.Tests/ConfigureOptionsTests.cs:314]: `TryConfigureSettings` is the "act"

File: src/Cli.Tests/ConfigureOptionsTests.cs (hunk 12 above)
Comment (seantleonard) [reply src/Cli.Tests/ConfigureOptionsTests.cs:286]: this is the case for most of these tests. please re-check all of them

File: src/Cli.Tests/EndToEndTests.cs [hunk 14]
```diff
@@ -1107,4 +1107,32 @@ public void TestRestRequestBodyStrictMode(bool includeRestRequestBodyStrictFlag,
         Assert.IsNotNull(runtimeConfig.Runtime.Rest);
         Assert.AreEqual(isRequestBodyStrict, runtimeConfig.Runtime.Rest.RequestBodyStrict);
     }
+
+    /// <summary>
+    /// Tests that the CLI can update the database type in the configuration file and is case insensitive.
```
Comment (seantleonard) [inline src/Cli.Tests/EndToEndTests.cs:1112]: sample command? dab configure ...

**Review by seantleonard — CHANGES_REQUESTED** [review]
Added some suggestions and clarification on tests. also some clarifying questions.

File: src/Cli.Tests/ConfigureOptionsTests.cs (hunk 10 above)
Comment (abhishekkumams) [reply src/Cli.Tests/ConfigureOptionsTests.cs:220]: it's the default value, it will give false if the key is not found.

File: src/Cli.Tests/ConfigureOptionsTests.cs
Comment (abhishekkumams) [reply src/Cli.Tests/ConfigureOptionsTests.cs:1]: they can do it one command as well. They're not mandated to do it in 2 commands. It's just we give them the flexibility that if they don't want to put everything in one command and choose to break it, we support it.

File: src/Cli/ConfigGenerator.cs (hunk 4 above)
Comment (abhishekkumams) [reply src/Cli/ConfigGenerator.cs:589]: No, it won't be. Because if the user first changes the database, that change is persisted. so the dbType will be cosmosDB when the second command will be run.

File: src/Cli/ConfigGenerator.cs (hunk 5 above)
Comment (abhishekkumams) [reply src/Cli/ConfigGenerator.cs:607]: sure.

File: src/Cli/ConfigGenerator.cs (hunk 6 above)
Comment (abhishekkumams) [reply src/Cli/ConfigGenerator.cs:15]: it's for using methods from CollectionUtils.
alt text

File: src/Cli.Tests/ConfigureOptionsTests.cs (hunk 9 above)
Comment (abhishekkumams) [reply src/Cli.Tests/ConfigureOptionsTests.cs:200]: yes, it overwrites the pre-existing one.

File: src/Cli/Commands/ConfigureOptions.cs (hunk 1 above)
Comment (sajeetharan) [reply src/Cli/Commands/ConfigureOptions.cs:54]: Should be Azure Cosmos DB for NOSQL (full caps)

Comment (abhishekkumams) [reply src/Cli/Commands/ConfigureOptions.cs:54]: why full caps for NoSQL? for example:
image
image

Also i feel fullcaps is not that readable.

File: src/Cli.Tests/ConfigureOptionsTests.cs
Comment (seantleonard) [reply src/Cli.Tests/ConfigureOptionsTests.cs:1]: > since using all the options in one command can be very long and errornous.
> It's just we give them the flexibility that if they don't want to put everything in one command and choose to break it, we support it.

would this reasoning be motivation for us to include Options short names? that way, developers can do something like:
`dab configure -dbname "myCosmosDbName" -dbType "CosmosDb" -container "cosmosContainer" -graphQLSchema "schema.gql"`

That way options are memorable and mitigate super verbose cmd line usage.

**Review by seantleonard — APPROVED** [review]

File: src/Cli.Tests/ConfigureOptionsTests.cs [hunk 15]
```diff
@@ -150,5 +150,211 @@ public void TestUpdateDepthLimitForGraphQL(int? newDepthLimit)
...
+        /// This method verifies that the database type can be changed from CosmosDB_NoSQL to MSSQL and that the 
+        /// specific MSSQL option 'set-session-context' is correctly added to the configuration and the specific
+        /// cosmosDB options are removed.
+        /// Command: dab configure --data-source.database-type mssql --data-source.options.set-session-context true 
+        /// </summary>
+        [TestMethod]
+        public void TestDatabaseTypeUpdateCosmosDB_NoSQLToMSSQL()
+        {
+            // Arrange
+            _fileSystem!.AddFile(TEST_RUNTIME_CONFIG_FILE, new MockFileData(INITIAL_COSMOSDB_NOSQL_CONFIG));
```
Comment (aaronburtle) [inline src/Cli.Tests/ConfigureOptionsTests.cs:204]: Looks like shared code between the //arrange steps in these tests, could refactor this out to eliminate the duplication.

**Comment by aaronburtle** [comment]
Question with respect to breaking the command up

`1. dab configure --data-source.database-type cosmosdb_nosql`
`2. dab configure --data-source.options.database testdbname --data-source.options.schema testschema.gql`

what happens if we put another command in between these 2, will it still accept the final command? If so, what happens if the command in between them changes the database-type to something other than cosmos?

so something like

`1. dab configure --data-source.database-type cosmosdb_nosql`
`2. dab configure --data-source.database-type mssql`
`3. dab configure --data-source.options.database testdbname --data-source.options.schema testschema.gql`



And then another case. If the 2nd command also included the database-type command

`1. dab configure --data-source.database-type cosmosdb_nosql`
`2.dab configure --data-source.database-type mssql --data-source.options.database testdbname --data-source.options.schema testschema.gql`

I presume that all of the above would fail but just want to double check.

**Comment by aaronburtle** [comment]
> And then another case. If the 2nd command also included the database-type command
> 
> `1. dab configure --data-source.database-type cosmosdb_nosql` 
> `2.dab configure --data-source.database-type mssql --data-source.options.database testdbname --data-source.options.schema testschema.gql`
> 
> I presume that all of the above would fail but just want to double check.


Follow up question, for the case that I quoted. Does order of the options within the command itself here matter? So for example what if we send these 2 commands, and note that the changing of the dbtype to mssql happens at the end of the 2nd command


`1. dab configure --data-source.database-type cosmosdb_nosql` 
`2.dab configure --data-source.options.database testdbname --data-source.options.schema testschema.gql` --data-source.database-type mssql 


Does this fail?

**Review by aaronburtle — APPROVED** [review]
LGTM! Just a comment about some code duplication in the tests, and then a question about some edge cases to make sure they get handled.

**Comment by abhishekkumams** [comment]
> Question with respect to breaking the command up
> 
> `1. dab configure --data-source.database-type cosmosdb_nosql` `2. dab configure --data-source.options.database testdbname --data-source.options.schema testschema.gql`
> 
> what happens if we put another command in between these 2, will it still accept the final command? If so, what happens if the command in between them changes the database-type to something other than cosmos?
> 
> so something like
> 
> `1. dab configure --data-source.database-type cosmosdb_nosql` `2. dab configure --data-source.database-type mssql` `3. dab configure --data-source.options.database testdbname --data-source.options.schema testschema.gql`
> 
> And then another case. If the 2nd command also included the database-type command
> 
> `1. dab configure --data-source.database-type cosmosdb_nosql` `2.dab configure --data-source.database-type mssql --data-source.options.database testdbname --data-source.options.schema testschema.gql`
> 
> I presume that all of the above would fail but just want to double check.

Yes, both the commands will fail. Every command gets checked for correct database-type. if that's changed in midway. it will error it out.

**Comment by abhishekkumams** [comment]
> > And then another case. If the 2nd command also included the database-type command
> > `1. dab configure --data-source.database-type cosmosdb_nosql`
> > `2.dab configure --data-source.database-type mssql --data-source.options.database testdbname --data-source.options.schema testschema.gql`
> > I presume that all of the above would fail but just want to double check.
> 
> Follow up question, for the case that I quoted. Does order of the options within the command itself here matter? So for example what if we send these 2 commands, and note that the changing of the dbtype to mssql happens at the end of the 2nd command
> 
> `1. dab configure --data-source.database-type cosmosdb_nosql` `2.dab configure --data-source.options.database testdbname --data-source.options.schema testschema.gql --data-source.database-type mssql`
> 
> Does this fail?

yes, because the options provided do not belong to mssql type.
